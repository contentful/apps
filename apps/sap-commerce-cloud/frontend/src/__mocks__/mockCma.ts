const mockCma: any = {
  appSignedRequest: {
    create: () => ({}),
  },
  getSpace: () => ({
    getEnvironment: () => ({
      getEditorInterfaces: () => ({
        items: [],
      }),
      getContentTypes: () => ({
        items: [],
      }),
    }),
  }),
};

export { mockCma };
