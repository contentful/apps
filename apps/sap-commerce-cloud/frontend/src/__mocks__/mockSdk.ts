import { vi } from 'vitest';

export const makeSdkMock = () => ({
  ids: {
    app: 'some-app',
    space: '123214',
    environment: 'master',
  },
  hostnames: {
    webapp: 'app.contentful.com',
  },
  window: {
    startAutoResizer: vi.fn(),
  },
  field: {
    type: 'Array',
    removeValue: vi.fn(),
    onValueChanged: vi.fn((value) => value),
    onIsDisabledChanged: vi.fn((value) => value),
    getValue: vi.fn(),
    id: 'fieldId',
    setValue: vi.fn((value) => {
      return value;
    }),
  },
  space: {
    sys: {
      id: 'spaceId',
    },
    getContentTypes: vi.fn().mockReturnValue({
      items: [
        {
          sys: { id: 'some-id' },
          name: 'some-name',
          fields: [
            {
              id: 'some-id',
              name: 'some-name',
              type: 'some-type',
            },
          ],
        },
      ],
    }),
    getEditorInterfaces: vi.fn().mockReturnValue({
      items: [
        {
          sys: { contentType: { sys: { id: 'some-id' } } },
        },
      ],
    }),
  },
  dialogs: {
    openCurrentApp: vi.fn(() => {
      return ['sku1', 'sku2'];
    }),
  },
  parameters: {
    installation: {},
    invocation: {
      apiEndpoint: 'https://api.example.com',
      baseSites: 'site1,site2',
      fieldValue: [],
    },
  },
  cma: {},
  cmaAdapter: {},
  app: {
    onConfigure: vi.fn(),
    getParameters: vi.fn().mockReturnValueOnce({}),
    setReady: vi.fn(),
    getCurrentState: vi.fn(),
  },
});
