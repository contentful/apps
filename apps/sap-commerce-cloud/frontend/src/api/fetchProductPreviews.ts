import difference from 'lodash/difference';
import { Product, SAPParameters } from '../interfaces';
import { productTransformer } from './dataTransformers';
import { BaseAppSDK, CMAClient } from '@contentful/app-sdk';
import { isHAAEnabled } from '../helpers/isHAAEnabled';
import { toHAAParams } from '../helpers/toHAAParams';

const fetchHAAProductPreviews = async (
  skus: string[],
  parameters: SAPParameters,
  ids: BaseAppSDK['ids'],
  cma: CMAClient
): Promise<Product[]> => {
  const { response } = await cma.appActionCall.createWithResponse(
    toHAAParams('fetchProductPreview', ids),
    {
      parameters: {
        sapApiEndpoint: parameters.installation.apiEndpoint,
        apiKey: ids.app,
        skus: JSON.stringify(skus),
      },
    }
  );
  return JSON.parse(response.body).products;
};
export async function fetchProductPreviews(
  skus: string[],
  parameters: SAPParameters,
  ids: BaseAppSDK['ids'],
  cma: CMAClient
): Promise<Product[]> {
  if (!skus.length) {
    return [];
  }

  if (isHAAEnabled(ids)) {
    return fetchHAAProductPreviews(skus, parameters, ids, cma);
  }

  let totalResponse: any[] = [];
  let skuIds: string[] = [];
  let skuIdsToSkusMap: { [key: string]: string } = {};

  for (const sku of skus) {
    const skuId = sku.split('/products/').pop() as string;
    skuIds.push(skuId);
    skuIdsToSkusMap[skuId] = sku;

    const url = `${sku}?fields=code,name,summary,price(formattedValue,DEFAULT),images(galleryIndex,FULL),averageRating,stock(DEFAULT),description,availableForPickup,url,numberOfReviews,manufacturer,categories(FULL),priceRange,multidimensional,configuratorType,configurable,tags`;
    const response = await fetch(url);
    if (response.ok) {
      let responseJson = await response.json();
      totalResponse.push(responseJson);
    }
  }

  const products = totalResponse.map(productTransformer(parameters.installation, skuIdsToSkusMap));
  const foundSKUs = products.map((product: { sku: any }) => product.sku);
  const missingProducts = difference(skuIds, foundSKUs).map((sku) => ({
    sku: sku.split('/products/').pop(),
    image: '',
    id: '',
    name: '',
    isMissing: true,
    productUrl: skuIdsToSkusMap[sku],
  }));
  return [...products, ...missingProducts];
}
