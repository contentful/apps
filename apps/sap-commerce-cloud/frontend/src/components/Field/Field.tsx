import { FieldAppSDK } from '@contentful/app-sdk';
import get from 'lodash/get';
import { SortableComponent } from '@components/ProductPreviews/SortableComponent';
import logo from '../../logo.png';
import { SAPParameters } from '@interfaces';
import { styles } from '@components/Field/Field.styles';
import { ShoppingCartIcon } from '@contentful/f36-icons';
import useAPI from '@hooks/useAPI';
import { useEffect, useState } from 'react';
import { IconButton } from '@contentful/f36-components';
import { useSDK } from '@contentful/react-apps-toolkit';

function fieldValueToState(value?: string | string[]): string[] {
  if (!value) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

function makeCTAText(fieldType: string) {
  const isArray = fieldType === 'Array';
  const beingSelected = isArray ? 'products' : 'a product';
  return `Select ${beingSelected}`;
}

export default function Field() {
  const sdk = useSDK<FieldAppSDK>();
  const [value, setValue] = useState<string[]>(fieldValueToState(sdk.field.getValue()));
  const [editingDisabled, setEditingDisabled] = useState<boolean>(true);
  const sapAPI = useAPI(sdk.parameters as unknown as SAPParameters, sdk.ids, sdk.cma);

  useEffect(() => {
    sdk.window.startAutoResizer();

    // Handle external changes (e.g. when multiple authors are working on the same entry).
    sdk.field.onValueChanged((newValue?: string[] | string) => {
      setValue(fieldValueToState(newValue));
    });

    // Disable editing (e.g. when field is not editable due to R&P).
    sdk.field.onIsDisabledChanged((isDisabled: boolean) => {
      setEditingDisabled(isDisabled);
    });
  }, [sdk]);

  const updateStateValue = (skus: string[]) => {
    setValue(skus);

    if (skus.length > 0) {
      const fieldValue = sdk.field.type === 'Array' ? skus : skus[0];
      sdk.field.setValue(fieldValue);
    } else {
      sdk.field.removeValue();
    }
  };

  const onDialogOpen = async () => {
    const skus = await sdk.dialogs.openCurrentApp({
      allowHeightOverflow: true,
      position: 'center',
      title: makeCTAText(sdk.field.type),
      shouldCloseOnOverlayClick: true,
      shouldCloseOnEscapePress: true,
      parameters: {
        ...sdk.parameters.installation,
        fieldValue: fieldValueToState(sdk.field.getValue()),
        fieldType: sdk.field.type,
        fieldId: sdk.field.id,
      },
      width: 1400,
      minHeight: 600,
    });

    const result = Array.isArray(skus) ? skus : [];

    if (result.length) {
      updateStateValue(result);
    }
  };

  const hasItems = value.length > 0;
  const fieldType = get(sdk, ['field', 'type'], '');

  return (
    <>
      {hasItems && (
        <div className={styles.sortable}>
          <SortableComponent
            sdk={sdk}
            disabled={editingDisabled}
            skus={value}
            onChange={updateStateValue}
            fetchProductPreviews={sapAPI.fetchProductPreviews}
          />
        </div>
      )}
      <div className={styles.container}>
        <img src={logo} alt="Logo" className={styles.logo} />
        <IconButton
          variant="secondary"
          icon={<ShoppingCartIcon size="small" variant="muted" />}
          onClick={onDialogOpen}
          aria-label={makeCTAText(fieldType)}
          isDisabled={editingDisabled}>
          {makeCTAText(fieldType)}
        </IconButton>
      </div>
    </>
  );
}
