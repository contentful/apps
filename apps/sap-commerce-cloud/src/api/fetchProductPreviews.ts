import difference from 'lodash/difference';
import { Product, SAPParameters } from '../interfaces';
import { productTransformer } from './dataTransformers';
import { config } from '../config';

export async function fetchProductPreviews(
  skus: string[],
  parameters: SAPParameters,
  applicationInterfaceKey: string
): Promise<Product[]> {
  if (!skus.length) {
    return [];
  }

  let totalResponse: any[] = [];
  let skuIds: string[] = [];

  const headers = config.isTestEnv
    ? {}
    : {
        headers: {
          'Application-Interface-Key': applicationInterfaceKey,
        },
      };
  for (const sku of skus) {
    const splitSku = sku.split(':');
    const baseSite = splitSku[0];
    const skuId = splitSku[1];
    skuIds.push(skuId);
    const response = await fetch(
      parameters.installation.apiEndpoint +
        `/occ/v2/${baseSite}/products/` +
        skuId +
        '?fields=code,name,summary,price(formattedValue,DEFAULT),images(galleryIndex,FULL),averageRating,stock(DEFAULT),description,availableForPickup,url,numberOfReviews,manufacturer,categories(FULL),priceRange,multidimensional,configuratorType,configurable,tags',
      headers
    );
    if (response.ok) {
      let responseJson = await response.json();
      totalResponse.push(responseJson);
    }
  }

  const products = totalResponse.map(productTransformer(parameters.installation));
  const foundSKUs = products.map((product: { sku: any }) => product.sku);
  const missingProducts = difference(skuIds, foundSKUs).map((sku) => ({
    sku,
    image: '',
    id: '',
    name: '',
    isMissing: true,
  }));
  return [...products, ...missingProducts];
}
