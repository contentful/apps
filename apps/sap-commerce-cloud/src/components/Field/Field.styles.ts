import { css } from 'emotion';
import tokens from '@contentful/forma-36-tokens';

export const styles = {
  sortable: css({
    marginBottom: tokens.spacingM,
  }),
  container: css({
    display: 'flex',
  }),
  logo: css({
    display: 'block',
    height: '30px',
    marginRight: tokens.spacingM,
  }),
};
