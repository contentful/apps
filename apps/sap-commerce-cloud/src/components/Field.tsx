import React from 'react';
import { Button } from '@contentful/forma-36-react-components';
import { FieldAppSDK } from '@contentful/app-sdk';
import { css } from 'emotion';
import tokens from '@contentful/forma-36-tokens';
import get from 'lodash/get';
import { SortableComponent } from './ProductPreviews/SortableComponent';
import { fetchProductPreviews } from '../api/fetchProductPreviews';
import logo from '../logo.png';
import { AppParameters, SAPParameters } from '../interfaces';

interface Props {
  sdk: FieldAppSDK<AppParameters>;
}

interface State {
  value: string[];
  editingDisabled: boolean;
}

const styles = {
  sortable: css({
    marginBottom: tokens.spacingM,
  }),
  container: css({
    display: 'flex',
  }),
  logo: css({
    display: 'block',
    // width: "30px",
    height: '30px',
    marginRight: tokens.spacingM,
  }),
};

function fieldValueToState(value?: string | string[]): string[] {
  if (!value) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

function makeCTAText(fieldType: string) {
  const isArray = fieldType === 'Array';
  const beingSelected = isArray ? 'products' : 'a product';
  return `Select ${beingSelected}`;
}

export default class Field extends React.Component<Props, State> {
  state = {
    value: fieldValueToState(this.props.sdk.field.getValue()),
    editingDisabled: true,
  };

  componentDidMount() {
    this.props.sdk.window.startAutoResizer();

    // Handle external changes (e.g. when multiple authors are working on the same entry).
    this.props.sdk.field.onValueChanged((value?: string[] | string) => {
      this.setState({ value: fieldValueToState(value) });
    });

    // Disable editing (e.g. when field is not editable due to R&P).
    this.props.sdk.field.onIsDisabledChanged((editingDisabled: boolean) => {
      this.setState({ editingDisabled });
    });
  }

  updateStateValue = (skus: string[]) => {
    this.setState({ value: skus });

    if (skus.length > 0) {
      const value = this.props.sdk.field.type === 'Array' ? skus : skus[0];
      this.props.sdk.field.setValue(value);
    } else {
      this.props.sdk.field.removeValue();
    }
  };

  onDialogOpen = async () => {
    const { sdk } = this.props;

    const skus = await sdk.dialogs.openCurrentApp({
      allowHeightOverflow: true,
      position: 'center',
      title: makeCTAText(sdk.field.type),
      shouldCloseOnOverlayClick: true,
      shouldCloseOnEscapePress: true,
      parameters: {
        ...sdk.parameters.installation,
        fieldValue: fieldValueToState(sdk.field.getValue()),
        fieldType: sdk.field.type,
        fieldId: sdk.field.id,
      },
      width: 1400,
      minHeight: 600,
    });
    const result = Array.isArray(skus) ? skus : [];

    if (result.length) {
      this.updateStateValue(result);
    }
  };

  render = () => {
    const { value: data, editingDisabled } = this.state;

    const hasItems = data.length > 0;
    const fieldType = get(this.props, ['sdk', 'field', 'type'], '');

    return (
      <>
        {hasItems && (
          <div className={styles.sortable}>
            <SortableComponent
              sdk={this.props.sdk}
              disabled={editingDisabled}
              skus={data}
              onChange={this.updateStateValue}
              fetchProductPreviews={(skus) =>
                fetchProductPreviews(skus, this.props.sdk.parameters as SAPParameters)
              }
            />
          </div>
        )}
        <div className={styles.container}>
          <img src={logo} alt="Logo" className={styles.logo} />
          <Button
            icon="ShoppingCart"
            buttonType="muted"
            size="small"
            onClick={this.onDialogOpen}
            disabled={editingDisabled}>
            {makeCTAText(fieldType)}
          </Button>
        </div>
      </>
    );
  };
}
