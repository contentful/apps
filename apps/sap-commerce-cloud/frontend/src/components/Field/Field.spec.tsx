import { mockCma } from '@__mocks__/mockCma';
import { makeSdkMock } from '@__mocks__/mockSdk';
import { fetchProductPreviews } from '@api/fetchProductPreviews';
import Field from '@components/Field/Field';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { vi } from 'vitest';

const sdk = makeSdkMock();

vi.mock('@contentful/react-apps-toolkit', () => ({
  useSDK: () => sdk,
  useCMA: () => mockCma,
}));

describe('Field Component', () => {
  afterEach(() => {
    vi.clearAllMocks();
    cleanup();
  });

  it('should render without crashing', () => {
    const { container } = render(<Field />);
    expect(container).toMatchSnapshot();
  });

  it('should start auto resizer on mount', () => {
    render(<Field />);
    expect(sdk.window.startAutoResizer).toHaveBeenCalled();
  });

  it('should update value on sdk.field.onValueChanged', () => {
    const { queryByAltText } = render(<Field />);
    const newValue = ['sku1', 'sku2'];

    sdk.field.onValueChanged.mock.calls[0][0](newValue);

    expect(queryByAltText('Logo')).toBeInTheDocument();
    expect(sdk.field.getValue).toHaveBeenCalled();
  });

  it('should fetch product previews and call setValue correctly', () => {
    const { container } = render(<Field />);
    const skus = ['sku1', 'sku2'];

    const sortableComponent = container.querySelector('.sortable');
    if (sortableComponent) {
      fireEvent.change(sortableComponent, { target: { value: skus } });

      expect(fetchProductPreviews).toHaveBeenCalledWith(skus, sdk.parameters);
    }
  });

  it('should update value on sdk.field.setValue', () => {
    const { container } = render(<Field />);
    const skus = ['sku1', 'sku2'];

    const sortableComponent = container.querySelector('.sortable');
    if (sortableComponent) {
      fireEvent.change(sortableComponent, { target: { value: skus } });

      expect(sdk.field.setValue).toHaveBeenCalledWith(skus);
    }
  });

  it('should remove value on sdk.field.removeValue', () => {
    const { container } = render(<Field />);
    const skus: any = [];

    const sortableComponent = container.querySelector('.sortable');
    if (sortableComponent) {
      fireEvent.change(sortableComponent, { target: { value: skus } });

      expect(sdk.field.removeValue).toHaveBeenCalled();
    }
  });

  it('should call sdk.field.onIsDisabledChanged', () => {
    render(<Field />);
    expect(sdk.field.onIsDisabledChanged).toHaveBeenCalled();
  });
});
