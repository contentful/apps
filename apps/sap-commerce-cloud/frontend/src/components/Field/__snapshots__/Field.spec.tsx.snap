// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Field Component > should render without crashing 1`] = `
<div>
  <div
    class="css-k008qs"
  >
    <img
      alt="Logo"
      class="css-do5lj7"
      src="/src/logo.png"
    />
    <button
      aria-label="Select products"
      class="css-ketnjv"
      data-test-id="cf-ui-icon-button"
      disabled=""
      type="button"
    >
      <span
        class="css-3u52eb"
      >
        <svg
          aria-hidden="true"
          class="css-1jrff5x"
          color="currentColor"
          data-test-id="cf-ui-icon"
          role="img"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49A1.003 1.003 0 0020 4H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
      <span
        class="css-40g50j"
      >
        Select products
      </span>
    </button>
  </div>
</div>
`;
